### PR TITLE
Bdtt 19 customers get notified of shipment status

### PR DESCRIPTION
**Pull request by rohanmk99**
# Description
https://bottlenose.atlassian.net/jira/software/projects/BDTT/boards/1?selectedIssue=BDTT-19

Upon Clicking Subscribe in ViewAllShipments page the following happens:
addsubscription is called in the complex scrapper via method in shipment.jsx
This calls simple microservice subscription.py (add method) Which checks if subscription already exist, if it doesnt then adds new row into subscription and the current status
image

Then in complex scrapper there is a function called sendsms. That calls getsubscription in subscription.py to get all active subscriptions. Then we go through the list and obtain the container number. Using the container number we call the scrapper to obtain status. Then we check if new status is different from original status when subscribed. If status is different we obtain phone number of user from user.py getnumber function. Then we send sms using twilio api. New status is updated in the table. so only if status changes again then sms will be sent
image

image

Then we have scheduler.py that calls the sendsms every 24 hours and runs in the background to check if status has changed. 
image

Then when we click unsubscribe we call the delete subscription method via shipment.jsx and then complex microservice deletesubscription that finally invokes simple microservice subscription. This removes the record from the subscription table.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Migration

# Checklist:

- [ ] I have added the appropriate labels to my PR
- [ ] My code follows the style guidelines of this project
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by rohanmk99** [comment]
reopen

File: services/backend/core/complex_scraper.py [hunk 1]
```diff
@@ -37,6 +39,139 @@
...
+    
+        return response
+@app.route("/sendsms", methods=['POST'])
+def sendsms():
+    response = invoke_http2("core_subscription", "subscription/getsubscriptions",prod, method='POST')
+    for i in range(len(response)):
+        container=response[i]["container_id"]
+        userid=response[i]["wguser_id"]
+        subscription_status=response[i]["status"]
+        shipment_type="ctr"
```
Comment (pandaponn) [inline services/backend/core/complex_scraper.py:97]: Why is shipment_type and directions hardcoded here? Wouldn't that mean if the shipment is a BL or EXPORT, then the SMS won't send?

File: services/backend/core/complex_scraper.py [hunk 2]
```diff
@@ -37,6 +39,139 @@
...
+            }
+            phone_number=invoke_http2("core_user", "user/getnumber",prod, method='POST', json=useriddata)
+            
+            account_sid = "AC7a7b489784baef97d21697b086b468ec"
+            auth_token = "eed37e9b4744d26ed34a67d4d3efe232"
+            client = Client(account_sid, auth_token)
+            message = client.messages.create(
+                body="There has been a status update in regards to your Container Number: "+identifier+" Status:"+response2,
+                from_="+15674004435",
+                to = "+65"+str(phone_number)
```
Comment (pandaponn) [inline services/backend/core/complex_scraper.py:165]: Can you verify my number so that I can test the Twilio notification on my phone as well?

**Review by pandaponn — CHANGES_REQUESTED** [review]
Good work on getting the notification feature to work. I have not been able to test it personally as my phone number isn't verified but based on the code, it seems like it would work as intended. 

Additionally, I think YMLU scraper does not work as the scraper is not returning status.

Also, could you implement some sort of visual response in the UI when user subscribes/unsubscribes to a shipment? (Similar to what Charmaine has for logging in and logging out) And disable the unsubscribe button if user is not subscribed, vice-versa.

Would also be good if users can subscribe/unsubscribe from the /status page and not only through the dashboard. (But this is not crucial)

Please help to make the changes ASAP as we are short on time and I also need to integrate this ASAP, otherwise good job 👍🏻

File: services/frontend/app/package.json [hunk 3]
```diff
@@ -1,83 +1,83 @@
...
-    "react-scripts": "5.0.1",
-    "react-spinners": "^0.13.8",
-    "react-toastify": "^9.1.1",
-    "sass": "^1.58.3",
-    "selenium": "^2.20.0",
-    "styled-components": "^5.3.6",
-    "web-vitals": "^2.1.4"
-  },
-  "scripts": {
-    "start": "PORT=80 react-scripts start",
```
Comment (charmjj) [inline services/frontend/app/package.json:42]: be careful of what you commit. put back this PORT=80

File: services/frontend/app/src/components/ViewShipments/ViewShipmentComponent.jsx [hunk 4]
```diff
@@ -45,16 +140,22 @@ export default function ViewShipmentComponent({ title, data, setLoading }) {
           {items.length === 0 && <div className="mx-1">No shipments found</div>}
           {items.length > 0 && items.map((item, index) => {
             return (
+              <div className="inline-flex">
```
Comment (charmjj) [inline services/frontend/app/src/components/ViewShipments/ViewShipmentComponent.jsx:143]: image

Shipment cards and your subscribe notification buttons are misaligned. Can you make sure that they are all of same length? Or you can try putting your subscribe/unsubscribe buttons into the card as well. Most importantly following Zhi Hao's point, we need some indication of whether the user has subscribed to notifications for the particular shipment or not

**Review by charmjj — CHANGES_REQUESTED** [review]
see comments

File: services/frontend/app/package.json (hunk 3 above)
Comment (JeffS97) [reply services/frontend/app/package.json:42]: Done

**Review by pandaponn — APPROVED** [review]
LGTM

**Review by charmjj — APPROVED** [review]
lgtm

**Pull request merged**